### PR TITLE
KFSPTS-25523 backport FINP-8341

### PR DESCRIPTION
**Pull request by db483**

### CODE REVIEW

**Review by cah292 — COMMENTED** [review]
Before I finish up this review, I wanted to clarify the end goals of this PR. I see that it's reintroducing the base code feature that auto-requeues documents in response to KIM role or responsibility changes, and that it reintroduces the similar functionality from the Account-Delegate-related docs (but with the FINP-8341 fix applied). Given those changes, are we going to continue using the SY-14-03 Document Requeuer Job like we are today?

If we intend to STOP using SY-14-03, then we need to make sure the functionals fully agree with us switching back to only using base code's requeue setup. (Parts of this PR's changes will affect the auto-requeues of ALL documents, not just those limited to the scope of the SY-14-03 job or the Account Delegate docs.) In addition, we need to ensure that the backport isn't pulling in unintended changes; I noticed that the overlay of AccountDelegateGlobalMaintainableImpl.displayMaxAccountDelegatesInfoMessage() is changing the names of the parameters and properties that are being used to generate the info message for the user.

If we intend to CONTINUE using SY-14-03, then I don't think the changes from this PR will address the unexpected-requeue issues. Future runs of SY-14-03 would still forcibly requeue the unintended documents, due to how we've set up the criteria for finding which documents to requeue. We may want to come up with an alternate solution, such as excluding more doc types from the Document Requeuer Job or restructuring the SQL used by that job.

Please let me know what our plans should be concerning the Document Requeuer Job, since that affects how we will move forward with fixing this issue.

**Comment by db483** [comment]
@cah292 Thank you for your questions. Cathy would like to test turning off the document requeuer batch job and having only this functionality in place for requeueing documents. Cathy is aware that this will only requeue the affected documents and not all documents.

When the review is finished please DO NOT merge the code to develop. Cathy would like to test this functionality in KFS-PREKEW in first instance.

Thank you.

File: src/main/java/edu/cornell/kfs/sys/batch/service/impl/CuSchedulerServiceImpl.java [hunk 1]
```diff
@@ -340,6 +341,49 @@ public void scheduleExceptionMessageJob(PersistedMessage message, String descrip
...
+        String jobName = "Delayed_Asynchronous_Call-" + Math.random();
+        
+        JobDataMap jobDataMap = new JobDataMap();
+        jobDataMap.put(MessageServiceExecutorJob.MESSAGE_KEY, message);
+        
+        JobBuilder jobBuilder = JobBuilder.newJob();
+        if (StringUtils.isNotBlank(description)) {
+            jobBuilder = jobBuilder.withDescription(description);
+        }
+        return jobBuilder.withIdentity(jobName, "Delayed_Asynchronous_Call")
```
Comment (cah292) [inline src/main/java/edu/cornell/kfs/sys/batch/service/impl/CuSchedulerServiceImpl.java:382]: There are several places in this class where the processing has been set up to behave differently for "Exception Messaging" jobs. Many of those areas may need adjustments to also have similar behavior for these delayed-async-call jobs. You might need to modify at least the following areas:

* The getJobIfPresent() method
* The 6-arg runJob() method
* The runJobWithoutQuartz() method

**Review by cah292 — CHANGES_REQUESTED** [review]
Changes look good overall, though I did find some areas that could use more cleanup.

I also noticed that this PR is pulling in other KualiCo changes to these files (such as cleanup of the logging statements to use parameterization). In general those extra changes look okay, but I did highlight one such spot in AccountDelegateGlobalMaintainableImpl that might need further adjustments.

File: src/main/resources/edu/cornell/kfs/sys/cu-spring-sys.xml [hunk 2]
```diff
@@ -388,6 +389,9 @@
         This bean is only meant for KSB-message-related processing, so we exclude it
         from the KFS-SYS module's "jobNames" property.
      -->
+    <bean id="delayedAsyncCallJob" parent="unscheduledJobDescriptor"
+          p:namespaceCode="KFS-SYS"
+          p:group="Delayed Asynconous Messaging"/>
```
[the remark this answers is not in the document]
Comment (cah292) [reply src/main/resources/edu/cornell/kfs/sys/cu-spring-sys.xml:394]: Also, this group might need to be named "Delayed_Asynchronous_Call" for consistency with how it's set up in the code.

File: src/main/java/org/kuali/kfs/coa/document/AccountDelegateGlobalMaintainableImpl.java [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+
+    /**
+     * Informational message about max account delegates
+     */
+    protected void displayMaxAccountDelegatesInfoMessage() {
+        String maxAccountDelegatesString = SpringContext.getBean(ParameterService.class).getParameterValueAsString(
+                AccountDelegateGlobal.class, COAParameterConstants.ACCOUNT_DELEGATES_LIMIT);
+        if (maxAccountDelegatesString != null && !maxAccountDelegatesString.isEmpty()) {
+            GlobalVariables.getMessageMap().putInfo(KFSConstants.GLOBAL_ERRORS,
+                    COAKeyConstants.INFO_DOCUMENT_DELEGATE_ACCOUNT_DELEGATES_LIMIT, maxAccountDelegatesString);
```
[the remark this answers is not in the document]
Comment (db483) [reply src/main/java/org/kuali/kfs/coa/document/AccountDelegateGlobalMaintainableImpl.java:75]: Good catch, looking into it.

[the remark this answers is not in the document]
Comment (db483) [reply src/main/java/org/kuali/kfs/coa/document/AccountDelegateGlobalMaintainableImpl.java:75]: Updated to use older constants

File: src/main/java/org/kuali/kfs/kim/impl/responsibility/ResponsibilityChangeQueueImpl.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.kuali.kfs.kew.impl.responsibility;
```
[the remark this answers is not in the document]
Comment (db483) [reply src/main/java/org/kuali/kfs/kim/impl/responsibility/ResponsibilityChangeQueueImpl.java:19]: updated

File: src/main/java/org/kuali/kfs/ksb/messaging/serviceproxies/DelayedAsynchronousServiceCallProxy.java [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+import java.io.Serializable;
+import java.lang.reflect.Method;
+import java.lang.reflect.Proxy;
+import java.sql.Timestamp;
+import java.util.Calendar;
+
+/**
+ * A proxy which schedules a service to be executed asynchronously after some delay period.
+ */
+public final class DelayedAsynchronousServiceCallProxy extends BaseInvocationHandler
```
[the remark this answers is not in the document]
Comment (db483) [reply src/main/java/org/kuali/kfs/ksb/messaging/serviceproxies/DelayedAsynchronousServiceCallProxy.java:56]: comment added

File: src/main/resources/edu/cornell/kfs/coa/document/datadictionary/AccountDelegateGlobalMaintenanceDocument.xml [hunk 6]
```diff
@@ -17,8 +17,8 @@
       
 
   <bean id="AccountDelegateGlobalMaintenanceDocument" parent="AccountDelegateGlobalMaintenanceDocument-parentBean">
-  <property name="businessObjectClass" value="edu.cornell.kfs.coa.businessobject.CuAccountDelegateGlobal"/>
-    <property name="maintainableClass" value="edu.cornell.kfs.coa.document.CuAccountDelegateGlobalMaintainableImpl"/>   
+  <property name="businessObjectClass" value="org.kuali.kfs.coa.businessobject.AccountDelegateGlobal"/>
```
[the remark this answers is not in the document]
Comment (db483) [reply src/main/resources/edu/cornell/kfs/coa/document/datadictionary/AccountDelegateGlobalMaintenanceDocument.xml:20]: updated

File: src/main/resources/edu/cornell/kfs/sys/cu-spring-sys.xml (hunk 2 above)
[the remark this answers is not in the document]
Comment (db483) [reply src/main/resources/edu/cornell/kfs/sys/cu-spring-sys.xml:394]: updated

File: src/main/resources/edu/cornell/kfs/sys/cu-spring-sys.xml [hunk 7]
```diff
@@ -388,6 +389,9 @@
         This bean is only meant for KSB-message-related processing, so we exclude it
```
[the remark this answers is not in the document]
Comment (db483) [reply src/main/resources/edu/cornell/kfs/sys/cu-spring-sys.xml:389]: updated

File: src/main/java/edu/cornell/kfs/sys/batch/service/impl/CuSchedulerServiceImpl.java (hunk 1 above)
Comment (db483) [reply src/main/java/edu/cornell/kfs/sys/batch/service/impl/CuSchedulerServiceImpl.java:382]: updated

**Comment by db483** [comment]
This is ready for review again. Thank you.

File: src/main/resources/edu/cornell/kfs/sys/cu-spring-sys.xml [hunk 8]
```diff
@@ -379,15 +379,19 @@
...
           class="org.kuali.kfs.ksb.messaging.quartz.MessageServiceExecutorJobListener"/>
 
     <!--
-        This bean is only meant for KSB-message-related processing, so we exclude it
+        These beans are only meant for KSB-message-related processing, so we exclude it
         from the KFS-SYS module's "jobNames" property.
      -->
+    <bean id="delayedAsyncCallJob" parent="unscheduledJobDescriptor"
+          p:namespaceCode="KFS-SYS"
+          p:group="Delayed Asynchronous Call"/>
```
Comment (cah292) [inline src/main/resources/edu/cornell/kfs/sys/cu-spring-sys.xml:394]: If you end up making the group name changes I had suggested in the other comment, then this value should be updated to match.

**Review by cah292 — CHANGES_REQUESTED** [review]
I noticed a few more items that should be cleaned up, but otherwise the revisions look good.

File: src/main/java/edu/cornell/kfs/sys/CUKFSConstants.java [hunk 9]
```diff
@@ -201,6 +201,8 @@ public static final class WebappPaths {
     public static final String CU_ALLOW_LOCAL_BATCH_EXECUTION_KEY = "cu.allow.local.batch.execution";
     public static final String EXCEPTION_MESSAGING_GROUP = "Exception Messaging";
     public static final String EXCEPTION_MESSAGE_JOB_NAME_PREFIX = "Exception_Message_Job ";
+    public static final String DELAYED_ASYNCRONOUS_CALL_GROUP = "Delayed Asynchronous Call";
+    public static final String DELAYED_ASYNCRONOUS_CALL_JOB_NAME_PREFIX = "Delayed_Asynchronous_Call ";
```
[the remark this answers is not in the document]
Comment (db483) [reply src/main/java/edu/cornell/kfs/sys/CUKFSConstants.java:205]: updating

File: src/main/java/edu/cornell/kfs/sys/batch/service/impl/CuSchedulerServiceImpl.java [hunk 10]
```diff
@@ -295,6 +296,8 @@ public void runJob(String groupName, String jobName, int startStep, int stopStep
             String requestorEmailAddress) {
         if (StringUtils.equals(groupName, CUKFSConstants.EXCEPTION_MESSAGING_GROUP)) {
             throw new IllegalArgumentException("This method cannot be used to execute Exception Messaging jobs");
+        } else if (StringUtils.equals(groupName, CUKFSConstants.DELAYED_ASYNCRONOUS_CALL_GROUP)) {
+            throw new IllegalArgumentException("This method cannot be used to execute Delayed Asyncronous Call jobs");
```
[the remark this answers is not in the document]
Comment (db483) [reply src/main/java/edu/cornell/kfs/sys/batch/service/impl/CuSchedulerServiceImpl.java:300]: updated

File: src/main/java/edu/cornell/kfs/sys/batch/service/impl/CuSchedulerServiceImpl.java [hunk 11]
```diff
@@ -340,6 +343,49 @@ public void scheduleExceptionMessageJob(PersistedMessage message, String descrip
...
+            throw e;
+        } finally {
+            if (!jobWasScheduled) {
+                jobs.remove(jobKey);
+            }
+        }
+    }
+    
+    private JobDetail createDelayedAsyncCallJobDetail(PersistedMessage message, String description) {
+        String jobName = "Delayed_Asynchronous_Call-" + Math.random();
```
[the remark this answers is not in the document]
Comment (db483) [reply src/main/java/edu/cornell/kfs/sys/batch/service/impl/CuSchedulerServiceImpl.java:375]: updated

File: src/main/java/edu/cornell/kfs/sys/batch/service/impl/CuSchedulerServiceImpl.java [hunk 12]
```diff
@@ -340,6 +343,49 @@ public void scheduleExceptionMessageJob(PersistedMessage message, String descrip
...
+        
+        JobDataMap jobDataMap = new JobDataMap();
+        jobDataMap.put(MessageServiceExecutorJob.MESSAGE_KEY, message);
+        
+        JobBuilder jobBuilder = JobBuilder.newJob();
+        if (StringUtils.isNotBlank(description)) {
+            jobBuilder = jobBuilder.withDescription(description);
+        }
+        return jobBuilder.withIdentity(jobName, "Delayed Asynchronous Call")
```
[the remark this answers is not in the document]
Comment (db483) [reply src/main/java/edu/cornell/kfs/sys/batch/service/impl/CuSchedulerServiceImpl.java:384]: updated

File: src/main/java/edu/cornell/kfs/sys/batch/service/impl/CuSchedulerServiceImpl.java [hunk 13]
```diff
@@ -412,6 +458,13 @@ private void runJobWithoutQuartz(CuJobEntry jobEntry, CuJobExecutionContext exec
                     }
                     jobs.remove(jobKey);
                 }
+                if (StringUtils.equals(jobKey.getGroup(), CUKFSConstants.DELAYED_ASYNCRONOUS_CALL_GROUP)) {
+                    if (LOG.isDebugEnabled()) {
+                        LOG.debug("runJobWithoutQuartz, Removing temporary delayed asyncronous call job: "
```
[the remark this answers is not in the document]
Comment (db483) [reply src/main/java/edu/cornell/kfs/sys/batch/service/impl/CuSchedulerServiceImpl.java:463]: updated

File: src/main/resources/edu/cornell/kfs/sys/cu-spring-sys.xml (hunk 8 above)
Comment (db483) [reply src/main/resources/edu/cornell/kfs/sys/cu-spring-sys.xml:394]: updated

**Review by cah292 — COMMENTED** [review]
The revised changes look good. Are you ready for me to merge this yet, or does Cathy still need to do more testing of it in Pre-KEW first?

**Comment by db483** [comment]
Thank you Chad. This should NOT be merged. Cathy needs to validate in kfs-PreKEW. preKEW is currently built and deployed daily from this branch.

**Comment by db483** [comment]
A new PR was created based on the 11/17/21 upgraded code: https://github.com/CU-CommunityApps/cu-kfs/pull/1340

**Pull request closed without merging**